File: atcoder/tyama_atcoderbcu30C_TLE.rb
#!/usr/bin/ruby
n,m=gets.split.map(&:to_i)
h={}
a=m.times.map{gets.split.map(&:to_i)}
a.each{|x,y|h[[x,y]]=1}
r=0
(1..n).each{|i|(1..n).each{|j|
	r+=(1..n).count{|k|h.has_key?([i,k])&&h.has_key?([k,j])}**2
}}
p r

File: atcoder/tyama_atcoderbcu30C_TLE.py
#!/usr/bin/python
import sys
n,m=map(int,sys.stdin.readline().split())
h=set()
a=[tuple(map(int,sys.stdin.readline().split())) for _ in range(m)]
for e in a: h.add(e)
r=0
for i in range(1,n+1):	
	for j in range(1,n+1):
		r+=sum((i,k) in h and (k,j) in h for k in range(1,n+1))**2
print(r)
